src/analyses/ai.h

```diff
@@ -82,15 +82,22 @@ class ai_baset
     finalize();
   }
 
+  /// Accessing individual domains at particular domains
+  /// (without needing to know what kind of domain or history is used)
+  /// A pointer to a copy as the method should be const and
+  /// there are some non-trivial cases including merging domains, etc.
+
   /// Returns the abstract state before the given instruction
-  virtual const ai_domain_baset & abstract_state_before(
-    goto_programt::const_targett t) const = 0;
+  /// PRECONDITION(l is dereferenceable)
+  virtual std::unique_ptr<statet> abstract_state_before(locationt l) const = 0;
```

Comment:
What would be the use-case for it not returning anything?

Reply:
Why can't it just return a const ref?

Reply:
In this case it could and used to return a constant ref.  However if we are going to break the 1-to-1 link between location and domain (which is the whole point of this patch set) then we have to be able to support cases where the abstract interpreter may not have the domain and may have to compute it in this function.  For example : merging all of the different contexts that reach this point or dynamically recomputing the domain from the last merge point.  That's why the API needs to change.